components/ai_chat/core/browser/engine/oai_api_client.cc

```diff
@@ -0,0 +1,180 @@
+  base::flat_map<std::string, std::string> headers;
+  if (!model_config.api_key.empty()) {
+    headers.emplace("Authorization",
+                    base::StrCat({"Bearer ", model_config.api_key}));
+  }
+
+  if (is_sse_enabled) {
+    auto on_received = base::BindRepeating(&OAIAPIClient::OnQueryDataReceived,
+                                           weak_ptr_factory_.GetWeakPtr(),
+                                           std::move(data_received_callback));
+    auto on_complete = base::BindOnce(&OAIAPIClient::OnQueryCompleted,
+                                      weak_ptr_factory_.GetWeakPtr(),
+                                      std::move(completed_callback));
+
+    api_request_helper_->RequestSSE(net::HttpRequestHeaders::kPostMethod,
+                                    model_config.endpoint, request_body,
+                                    "application/json", std::move(on_received),
+                                    std::move(on_complete), headers, {});
+  } else {
+    auto on_complete = base::BindOnce(&OAIAPIClient::OnQueryCompleted,
+                                      weak_ptr_factory_.GetWeakPtr(),
+                                      std::move(completed_callback));
+    api_request_helper_->Request(
+        net::HttpRequestHeaders::kPostMethod, model_config.endpoint,
+        request_body, "application/json", std::move(on_complete), headers, {});
+  }
+}
+
+void OAIAPIClient::OnQueryCompleted(GenerationCompletedCallback callback,
+                                    APIRequestResult result) {
+  const bool success = result.Is2XXResponseCode();
+  // Handle successful request
+  if (success) {
+    std::string completion = "";
+    // We're checking for a value body in case for non-streaming API results.
+    if (result.value_body().is_dict()) {
+      const base::Value::List* choices =
+          result.value_body().GetDict().FindList("choices");
+      const base::Value::Dict* delta =
+          choices->front().GetDict().FindDict("message");
+      completion = *delta->FindString("content");
+    }
+
+    std::move(callback).Run(base::ok(std::move(completion)));
+    return;
+  }
+
+  // Handle error
+  mojom::APIError error;
+
+  if (net::HTTP_TOO_MANY_REQUESTS == result.response_code()) {
+    error = mojom::APIError::RateLimitReached;
+  } else if (net::HTTP_REQUEST_ENTITY_TOO_LARGE == result.response_code()) {
+    error = mojom::APIError::ContextLimitReached;
+  } else {
+    error = mojom::APIError::ConnectionIssue;
+  }
+
+  std::move(callback).Run(base::unexpected(std::move(error)));
+}
+
+void OAIAPIClient::OnQueryDataReceived(
+    GenerationDataCallback callback,
+    base::expected<base::Value, std::string> result) {
+  if (!result.has_value() || !result->is_dict()) {
+    return;
+  }
+
+  // This client only supports completion events
+  const base::Value::List* choices = result->GetDict().FindList("choices");
+  const base::Value::Dict* delta = choices->front().GetDict().FindDict("delta");
```

Comment:
Do we need to validate if it is a Dict (here and in `OnQueryCompleted`)?

Reply:
aren't we validating already and returning early? in this function `if (!result->is_dict()) {}`

Reply:
added dict checks